midend/actionSynthesis.cpp

```diff
@@ -125,6 +125,10 @@ const IR::Node* DoSynthesizeActions::postorder(IR::P4Control* control) {
 }
 
 const IR::Node* DoSynthesizeActions::preorder(IR::BlockStatement* statement) {
+    if (statement->annotations->getSingle("atomic")) {
```

Comment:
Nothing really promises that action bodies are atomic.
Perhaps the created action should also have an atomic annotation?

Reply:
Atomic must be honored by the backend as well.

Reply:
Good point, I see that `createAction` function does end up with creating an action with a `hidden` and the action body with the `atomic` annot.

I can update the action to also have the `atomic` annotation. Would it make sense to extend this to other annotations on the block as well?

Reply:
It would make sense to preserve the other annotations, but it is very difficult to make a blanket statement about annotations, since their semantics is not standardized (except the ones defined by the spec). Preserving them would not hurt.

Reply:
Thanks, I added a fix to copy the annotations to the newly created action.

Reply:
The question remains why move an atomic block to an action?
Can't an action itself have inside one or more atomic blocks?
Perhaps this should be guided by a policy property.

Reply:
I suspect that the reason is that there is an "oral tradition" of P4 implementations that treat action bodies as if they have @atomic around their bodies, as BMv2 does.  It is certainly nothing from the language spec that encourages this.

On the other hand, I would guess that statements like "it is very difficult to make a blanket statement about annotations" probably also discourage people from wanting to rely on the language-standard @atomic annotation for things that they care about whether they are preserved by the p4c front end.  If there are no rules or invariants that annotations follow that are well known among compiler developers, they become mysterious scary things that lead to compiler bugs.

Reply:
The compiler knows about atomic, so it should definitely handle that.
The compiler front-end does not know anything about other target-specific annotations that may exist.